src/Microsoft.AspNet.OData.Shared/Query/Expressions/FilterBinder.cs

```diff
@@ -329,16 +329,16 @@ private Expression BindSingleResourceCastFunctionCall(SingleResourceFunctionCall
                 targetClrType = EdmLibHelpers.GetClrType(targetEdmType.ToEdmTypeReference(false), Model);
             }
 
+            Expression source = arguments[0];
+
             if (arguments[0].Type == targetClrType)
             {
                 // We only support to cast Entity type to the same type now.
-                return arguments[0];
+                return source;
             }
             else if (arguments[0].Type.IsAssignableFrom(targetClrType))
             {
                 // To support to cast Entity/Complex type to the sub type now.
-                Expression source = BindCastSourceNode(node.Source);
```

Comment:
@KenitoInc Would you care to elaborate what the call to `BindCastSourceNode` was doing and how/why removing it fixes the issue? You could add that to the PR discussion...